handlers/oauthhandler/providers.go

```diff
@@ -0,0 +1,192 @@
+	Name          string
+	AuthURL       string
+	TokenURL      string
+	ProfileURL    string
+	ClientID      string
+	ClientSecret  string
+	Scope         string
+	UsernameField string
+}
+
+// OAuthToken is the OAuth token.
+type OAuthToken struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"`
+}
+
+// NewProvider creates a new OAuth provider.
+func NewProvider(name, authURL, tokenURL, profileURL, clientID, clientSecret, scope string, usernameField string) *Provider {
+	return &Provider{
+		Name:          name,
+		AuthURL:       authURL,
+		TokenURL:      tokenURL,
+		ProfileURL:    profileURL,
+		ClientID:      clientID,
+		ClientSecret:  clientSecret,
+		Scope:         scope,
+		UsernameField: usernameField,
+	}
+}
+
+func Init(pid string) (map[string]*Provider, error) {
+	// Load the .env file
+	err := godotenv.Load()
+	if err != nil {
+		log.Errorf("Error loading .env file")
+	}
+
+	// Read the configuration file.
+	filename := filepath.Join("handlers", "oauthhandler", "config.yml")
+	data, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read configuration file: %v", err)
+	}
+
+	// Parse the configuration file.
+	var cfg Config
+	err = yaml.Unmarshal(data, &cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse configuration file: %v", err)
+	}
+
+	// Initialize the providers.
+	providers := make(map[string]*Provider, len(cfg.Providers))
+	for name, conf := range cfg.Providers {
+		provider := NewProvider(
+			conf.Name,
+			conf.AuthURL,
+			conf.TokenURL,
+			conf.ProfileURL,
+			os.Getenv(conf.ClientID),
+			os.Getenv(conf.ClientSecret),
+			conf.Scope,
+			conf.UsernameField,
+		)
+		providers[name] = provider
+	}
+
+	return providers, nil
+}
+
+// GetAuthURL returns the OAuth authorize URL for the provider.
+func (p *Provider) GetAuthURL(redirectURL string) string {
+	u, _ := url.Parse(p.AuthURL)
+	q := u.Query()
+	q.Set("client_id", p.ClientID)
+	q.Set("redirect_uri", redirectURL)
+	q.Set("scope", p.Scope)
+	u.RawQuery = q.Encode()
+	return u.String()
+}
+
+// GetOAuthToken obtains the OAuth token for the provider using the authorization code.
+func (p *Provider) GetOAuthToken(code string, redirectURL string) (*OAuthToken, error) {
+	data := url.Values{}
+	data.Set("grant_type", "authorization_code")
+	data.Set("client_id", p.ClientID)
+	data.Set("client_secret", p.ClientSecret)
+	data.Set("redirect_uri", redirectURL)
+	data.Set("code", code)
+
+	req, err := http.NewRequest("POST", p.TokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			fmt.Printf("error closing HTTP body: %v\n", err)
```

Comment:
Actually, the best would be to use log.Errorw() or log.Warnw()

However, this is a not relevant error, since it might because of the client disconnecting. I'd then maybe only print it on Debug level to not spam the logs with unuseful messages.